apollo-federation/tests/subgraph/subgraph_validation_tests.rs

```diff
@@ -0,0 +1,300 @@
+use apollo_federation::error::FederationError;
+use apollo_federation::subgraph::SubgraphError;
+use apollo_federation::subgraph::typestate::Subgraph;
+
+fn build_for_errors(schema: &str, err_reason: &str) -> Result<Vec<Vec<String>>, SubgraphError> {
+    let err = Subgraph::parse("S", "", schema)
+        .expect("parses schema")
+        .expand_links()
+        .expect("expands links")
+        .validate(true)
+        .expect_err(err_reason);
+
+    match err.error {
+        FederationError::SingleFederationError(e) => {
+            let err_code_definition = e.code().definition();
+            let res = vec![vec![
+                err_code_definition.code().to_string(),
+                err_code_definition.doc_description().to_string(),
+            ]];
+            Ok(res)
+        }
+        FederationError::MultipleFederationErrors(e) => {
+            let mut res = Vec::new();
+            for error in e.errors {
+                let err_code_definition = error.code().definition();
```

Comment:
I'd like to avoid cloning code here, since we can just share the case for `SingleFederationError`.